### ISSUE TITLE
Add CIS benchmark agent for easy compliance reporting

### ISSUE THREAD

**Issue by stmcginnis** [bottlerocket-os/bottlerocket#2731]
**What I'd like:**

Bottlerocket has gone through [CIS benchmarking](https://github.com/bottlerocket-os/bottlerocket/issues/1297), but it was a manual process to collect all necessary data. Compliance can also vary based on configuration.

To make this easier to check for future certification refresh, and so users are able to easily get the status of their unique deployments, it would be useful to have an agent as part of Bottlerocket that could be triggered to perform a series of check and report the results.

Bottlerocket exposes some actions via the API such as `apiclient reboot`. It could be useful to have something like `apiclient cis-benchmark` or `cis-report` that would execute these checks and provide output that could be used to evaluate the compliance. There isn't a standard for this reporting format, but it may be useful to adhere to [localtoast](https://github.com/google/localtoast) or [OpenSCAP](http://www.open-scap.org/) formats for the sake of some consistency.

**Any alternatives you've considered:**

We can keep this a manual thing, but that is very error prone and very labor intensive to do on a regular basis.

**Comment by stmcginnis** [comment]
# CIS Agent Plan

## Overview

Bottlerocket has passed CIS Benchmarking. Compliance status is not a static thing, and Bottlerocket users need an easy way to tell if they are maintaining compliance.

AWS has published a [blog and a set of scripts](https://aws.amazon.com/blogs/containers/validating-amazon-eks-optimized-bottlerocket-ami-against-the-cis-benchmark/) that may be used in a host container for evaluating status, but this requires additional work by the end user to figure out how to set things up to run the scripts. The scripts referenced in the blog are not easy for the Bottlerocket maintainers to update as Bottlerocket features change or compliance items are updated. There also is not an easy way for an end user to use these checks to automate and enforce their compliance requirements.

## Requirements

An in-the-box CIS agent is being proposed that provides end users with easy access to a way to evaluate and enforce CIS benchmark compliance.

- Built-in agent that can evaluate system configuration and attest to compliance or non-compliance to the benchmark specification.
- Results and status accessible via the `apiclient`.
- Ability to do level 1 and/or level 2 compliance checks.
- Report output that can be programmatically parsed or imported into other tools.
- Report only on “Automated” checks.
- Note the “Manual” checks that are skipped to make it clear there are some items not completely covered by the agent automation.

### Out of scope

- Does not cover higher level things like Kubernetes CIS hardening. Established tools such as [KubeBench](https://github.com/aquasecurity/kube-bench) are better suited for that level of validation.
- Will not do auto-remediation of non-compliant checks, only reporting of compliance status.

## Options

There are a few open source CIS validation tools found on GitHub. Unfortunately, many of them appear to not have been touched for several years, and/or are implemented in something like Python that would cause challenges for use in Bottlerocket.

### Localtoast

Localtoast is one of the few recent tools on GitHub, and actually was created for validating Google’s Container-Optimized OS. This means there are a lot of similar requirements as we would have for integrating with Bottlerocket.

The tool is flexible in that it allows new checks to be defined using a protocol buffer based config file. This, at least in theory, would make it easy for us to define Bottlerocket-specific checks to perform, but it also makes the building and inclusion of Localtoast in the Bottlerocket image more difficult with build-time binary dependencies.

**Pros:**

- Extensible checks
- Existing framework
- Consistent reporting format to Google COS

**Cons:**

- Not clear if Bottlerocket needs would be acceptable to integrate
- Need for protocol buffer compiler (`protoc`) and `protoc-go-gen` plugin to build

### Custom Framework

Another option would be to build our own CIS agent framework. At its core, the agent functionality is a matter of checking for specific configuration or object state and reporting whether it matches the expectation in the benchmark. So there isn’t anything too complex about the checks performed. Bottlerocket just needs something that can execute those checks, collect and aggregate the results of the set of checks, and report the overall status.

**Pros:**

- Full control over what can be checked and how
- Can implement easy way to drop in new checks
- Ability to provide multiple/custom output formats
- Can be customized to exactly how Bottlerocket needs it
- Can be written in Rust for toolchain benefits (faster, smaller, safer)

**Cons:**

- Need to implement and maintain entire framework

## Recommendation

Options are limited to either adapting an existing tool (maintained by someone else), or implementing a custom tool.

While Localtoast has much of the functionality Bottlerocket would need, it also has some technical challenges for inclusion in Bottlerocket. And while collaboration with another project to solve a common challenge would could be a net benefit, it also brings complexity to adapting to Bottlerocket-specific needs. Therefore, my recommended approach for a CIS Benchmark Agent in Bottlerocket is to implement a bespoke solution.

## Implementation

Implement a new tool to be called `bloodhound` that can run a set of checks and collate their results.

```bash
$ bloodhound -h

Usage: bloodhound [OPTIONS] --format <FORMAT> --level <LEVEL> --output <OUTPUT>

Options:
  -c, --checks <CHECK_DIR>  Path to the directory containing checker binaries [default: /etc/bloodhound/checks]
  -f, --format <FORMAT>     Format of the output [default: text]
  -l, --level <LEVEL>       The CIS benchmark compliance level to check [default: 1]
  -o, --output <OUTPUT>     Write output to a file at given path [default: stdout]
  -h, --help                Print help
```

### Command line options

Option | Input | Description
-- | -- | --
--checks | [/path/to/checkers] | Provide the directory to discover available checks to run.
--format | text (default) | Print results output in human readable format.
json | Print results output as JSON.
--level | number | The CIS benchmark compliance level to check. Defaults to level 1.
--output | [/path/to/output] | If provided, write output to a file at given path. Otherwise print the results to stdout.

#### Checks

A command line way to point `bloodhound` at a directory containing checks to run.

#### Format

The default format will be human readable output. This will be similar to the example provided in the original CIS benchmark blog post:

```txt
[PASS] 1.1.1.1 Ensure mounting of udf filesystems is disabled (Automated)
[PASS] 1.3.1 Ensure dm-verity is configured (Automated)
[PASS] 1.4.1 Ensure setuid programs do not create core dumps (Automated)
```

The user may also specify `--format json` to have an easier format to use for programmatic consumption:

```json
{
  "level": 1,
  "total": 15,
  "passed": 11,
  "skipped": 4,
  "failed": 0,
  "status": "PASS",
  "timestamp": "2023-03-15T14:05:04.803284748Z",
  "results": {
    "br01020100": {
      "name": "br01020100",
      "id": "1.2.1",
      "level": 1,
      "title": "Ensure software update repositories are configured",
      "mode": "Manual",
      "status": "SKIP",
      "error": "Manual check, see benchmark for audit details."
    },
   { ... }
  }
}
```

#### Level

The user may specify to run checks for `--level 1` to run only the CIS Benchmark level 1 checks, or `--level 2` to run checks for both level 1 and level 2.

#### Output

If not specified, the output from running tests will be printed to stdout. Otherwise the user may provide a path to a file where the output will be written.

### ApiClient Integration

A typical user would interact with the benchmark agent via the `apiclient` command. Running the command `apiclient cisreport` would trigger the execution of the `bloodhound` command. Output would be output to the console from the `apiclient` command.

### High Level Architecture

The `bloodhound` executable itself will not contain any Bottlerocket-specific logic. It will be a generic testing harness responsible for loading the appropriate checks to run, executing each check, gathering check results, and reporting the status back.

The benchmark logic will be implemented in individual "checker" binaries. The main process will look in a designated directory for all available checks. It will then call in to each one to extract metadata about the check that will be used for determining things like benchmark level, name, and description. It can use this metadata to determine whether to include the check based on what the user has requested (e.g. `--level 2`).

Example of check subdirectory:

```txt
checks
├── bottlerocket-checks -> br-01010101*
├── bottlerocket-checks -> br-01030100*
├── bottlerocket-checks -> br-01040100*
├── bottlerocket-checks -> br-01040101*
├── bottlerocket-checks -> br-01040102*
└── br-01040103
```

Keeping the checks as independent binaries leaves the possibility for non-Bottlerocket users to potentially use `bloodhound` for their own validation checks. This may help encourage non-Bottlerocket collaboration.

There is some overhead having each individual checker be its own binary. As a compromise to reduce the overhead, while still allowing the simplicity of being able to drop new checkers into a designated directory, most Bottlerocket checks will be implemented in a multicall binary. In the example above, the `bottlerocket-checks` is actually one binary that contains multiple checks. Installation in Bottlerocket will create symlinks to this binary using different names. The `bottlerocket-checks` binary will inspect `args[0]` to see how it was called to determine which check should be executed.

This reduces the (small) overhead while keeping it simple for things like out of tree variants to easily drop in their own checks.

Keeping the benchmark agent as a harness and having the individual checks as separate binaries also enables non-Bottlerocket users to implement their checks in Rust, Go, Python, Bash, or whatever language makes the most sense for them and their target environment. However common helper code for formatting metadata and results will only be provided in Rust as part of the main `bloodhound` code base.

#### Checkers

The checker interface will have two mandatory calls:

```txt
$ [check] metadata
{name: "br-01010101", id: "1.1.1.1", level: 2, desc: "Ensure mounting of..."}

$ [check] # no argument
{status: "FAIL", error: "UDF kernel module is present."}
```

**Note:** the checkers could include more details such as remediation steps, rationale, versions, etc. But for the sake of image size, these should not be included in the checks themselves, but rather users should be referred to external documentation like the official CIS Benchmark report, or eventually the Bottlerocket project website, to get complete details for a failing check.

#### Execution

As mentioned, `bloodhound` itself would just be a harness for loading and running the checks. Its responsibilities would include:

- List files in `checks` directory to get set of checkers to parse
- Iterate through each check and call `metadata` on it to obtain check information
- Filter checks to be run based on input (e.g. `if metadata.level == 2 and args.level == 1) { skip }`)
- Iterate through each filtered check and execute, capturing results
- Collect composite results to get overall status, write formatted output to destination

Checker execution could be made parallel. This has the potential to cause conflicts between checkers and add unnecessary complexity to the code. So unless it is determined parallelism is desirable, all checks will be executed serially.

### Code Structure

Each checker to be added, as well as the`bloodhound` command binary, will be under its own separate path. Checkers will be under a separate directory to separate them from possible future checkers for non-Bottlerocket checkers. There will also be shared library code that the checkers can leverage for reporting metadata and results. As mentioned above, each Bottlerocket check will ultimately end up in one multicall binary that routes to the appropriate checker based on the name used to call the binary.

A new package will be added to all Bottlerocket variants pulling in releases from the bloodhound repo.

Code will be added to `apiclient` and `apiserver` to route `cisreport` calls through to execute `bloodhound` and report results.

> The bloodhound is a large scent hound, originally bred for hunting.

**Comment by stockholmux** [comment]
I really like this idea. One note: Since it seems like `bloodhound` could be used for other testing, would it make sense to make the the user-facing `apiclient` more universal as well?  Perhaps changing the endpoint from `cis-benchmark` to something more generic with an option - something like: `apiclient benchmark --check=cis`.

**Comment by stmcginnis** [comment]
> I really like this idea. One note: Since it seems like `bloodhound` could be used for other testing, would it make sense to make the the user-facing `apiclient` more universal as well? Perhaps changing the endpoint from `cis-benchmark` to something more generic with an option - something like: `apiclient benchmark --check=cis`.

I was on the fence between making it very generic like that, and leaving it to the user to pass the right args, versus being specific with some defaults. I went with the latter approach thinking if we needed to we could add something like `apiclient pci-compliance` as an example. That makes the user experience for this more clear and simpler. Internally both would then get routed to calling `bloodhound` with the right arguments for the given request.

Either way is valid, so good to hear feedback on this.
